### PR TITLE
Improve set instance variable

### PR DESCRIPTION
**Pull request by tenderlove**
This commit improves the set ivar implementation.

### CODE REVIEW

**Comment by maximecb** [comment]
Looking pretty good. Going to benchmark this on AWS.

@tenderlove could you add the test for writing embedded/extended?

**Comment by tenderlove** [comment]
@maximecb the test cases should be updated now

File: yjit_codegen.c [hunk 1]
```diff
@@ -744,6 +744,117 @@ enum {
...
+    // ID for the name of the ivar
+    ID id = ivar_name;
+    struct rb_iv_index_tbl_entry *ent;
+    struct st_table *iv_index_tbl = ROBJECT_IV_INDEX_TBL(comptime_receiver);
+
+    // Bail if this is a heap object, because this needs a write barrier
+    ADD_COMMENT(cb, "guard value is not frozen");
+    test(cb, REG1, imm_opnd(RUBY_IMMEDIATE_MASK));
+    jz_ptr(cb, COUNTED_EXIT(side_exit, setivar_val_heapobject));
+
```
Comment (XrXr) [inline yjit_codegen.c:777]:
```suggestion
```

I think this one is a duplicate

Comment (maximecb) [reply yjit_codegen.c:777]: Found the same bug 👍

File: bootstraptest/test_yjit.rb [hunk 2]
```diff
@@ -239,6 +239,86 @@ def itself
...
+    def write_one
+      @one = 1
+    end
+  end
+
+  foo = WillExtend.new
+  foo.write_one
+  foo.write_one
+  foo.make_extended
+  foo.write_one
```
[the remark this answers is not in the document]
Comment (tenderlove) [reply bootstraptest/test_yjit.rb:303]: Ok, I did that.

**Pull request merged**